### PR TITLE
Handle SIGTERM in Mobly.

### PR DESCRIPTION
**Pull request by booneng**
Handle SIGTERM in Mobly to allow Mobly to complete writing out summary report.

- - -
This change is [Reviewable](https://reviewable.io/reviews/google/mobly/858)

### CODE REVIEW

**Review by mhaoli — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/858)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @booneng and @xpconanfan)

___
*[`mobly/test_runner.py` line 78 at r1](https://reviewable.io/reviews/google/mobly/858#-NIkt0Sd3eYP2CsZ2Ca7:-NIkt0Sd3eYP2CsZ2Ca8:b-4ohlk2) ([raw file](https://github.com/google/mobly/blob/759f2c62b1c53835eeb5f0f19f3bd6bfbe26b093/mobly/test_runner.py#L78)):*
> ```Python
>   def sigterm_handler(*args):
>     raise signals.TestAbortAll(
>         'Test was terminated. This could be due to a timeout.')
> ```

As this is added to the public Mobly repo, so it may receive SIGTERM for many reasons. Not sure if it's better that we don't guess the reason, but just say we get a SIGTERM signal? 



_Code quote:_
```Python
This could be due to a timeout.
```

**Review by booneng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/858)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @mhaoli and @xpconanfan)

___
*[`mobly/test_runner.py` line 78 at r1](https://reviewable.io/reviews/google/mobly/858#-NIkt0Sd3eYP2CsZ2Ca7:-NIkyNvV6ZDSTC5DGH7I:b-896fix) ([raw file](https://github.com/google/mobly/blob/759f2c62b1c53835eeb5f0f19f3bd6bfbe26b093/mobly/test_runner.py#L78)):*
<details><summary><i>Previously, mhaoli (Minghao Li) wrote…</i></summary><blockquote>

As this is added to the public Mobly repo, so it may receive SIGTERM for many reasons. Not sure if it's better that we don't guess the reason, but just say we get a SIGTERM signal? 
</blockquote></details>

Done.

**Review by mhaoli — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/858)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @booneng and @xpconanfan)

___
*[`mobly/test_runner.py` line 414 at r2](https://reviewable.io/reviews/google/mobly/858#-NIl-AsJ7v6Akt6_W55U:-NIl-AsJ7v6Akt6_W55V:b-h9rfuc) ([raw file](https://github.com/google/mobly/blob/60512bdd00b7e1a6fa09845aabcf525649e1f8d0/mobly/test_runner.py#L414)):*
> ```Python
>     # finally blocks to be executed.
>     def sigterm_handler(*args):
>       raise signals.TestAbortAll('Test received a SIGTERM.')
> ```

Nit: seems that we can print a log before raising, we can tell users we get the SIGTERM but we will do cleanup before exit. This might be helpful bc users may expect we exit right now, but we may need some time.



_Code quote:_
```Python
raise signals.TestAbortAll('Test received a SIGTERM.')
```

File: mobly/test_runner.py [hunk 1]
```diff
@@ -75,12 +76,13 @@ def main(argv=None):
       runner.add_test_class(config, test_class, tests)
       try:
         runner.run()
-        ok = runner.results.is_all_pass and ok
```
Comment (xpconanfan) [inline mobly/test_runner.py:78]: why remove this one?

**Review by xpconanfan — CHANGES_REQUESTED** [review]

**Review by booneng — COMMENTED** [review]
If we have a scenario where the first test fails and the next test receives TestAbortAll, then the overall result will still be passed because the TestAbortAll is ignored in this block. So I moved this line to finally, so the correct result will be set when TestAbortAll is received.

> *[Reviewable](https://reviewable.io/reviews/google/mobly/858#-:-NIp5ap59ezmgzRYab1M:b-rqyioe)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @booneng and @mhaoli)

**Review by booneng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/858)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @booneng and @mhaoli)

___
*[`mobly/test_runner.py` line 414 at r2](https://reviewable.io/reviews/google/mobly/858#-NIl-AsJ7v6Akt6_W55U:-NIp7Qt4751pKLBlq4Xg:b-896fix) ([raw file](https://github.com/google/mobly/blob/60512bdd00b7e1a6fa09845aabcf525649e1f8d0/mobly/test_runner.py#L414)):*
<details><summary><i>Previously, mhaoli (Minghao Li) wrote…</i></summary><blockquote>

Nit: seems that we can print a log before raising, we can tell users we get the SIGTERM but we will do cleanup before exit. This might be helpful bc users may expect we exit right now, but we may need some time.
</blockquote></details>

Done.

File: mobly/test_runner.py (hunk 1 above)
Comment (xpconanfan) [reply mobly/test_runner.py:78]: Please reply comments in their original thread.

It sounds like in that case the test should pass, so it's expected behavior, no?

In any case, that's a different change; should be a separate PR if needed

**Review by booneng — COMMENTED** [review]
Ahh okay if that is the intended behaviour then that is fine.

I reverted this change.

> *[Reviewable](https://reviewable.io/reviews/google/mobly/858#-:-NIpH9be0Mw64g_rM0Yu:b-gl6vmr)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @booneng and @mhaoli)

File: mobly/test_runner.py (hunk 1 above)
Comment (booneng) [reply mobly/test_runner.py]: Reverted.

**Review by xpconanfan — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/858)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @booneng and @mhaoli)

___
*[`mobly/test_runner.py` line 407 at r5](https://reviewable.io/reviews/google/mobly/858#-NJ3wspm3FL4ZZVNAWla:-NJ3wspm3FL4ZZVNAWlb:bmxta1g) ([raw file](https://github.com/google/mobly/blob/7e7cfdc6a5706b65a620c9bcc44a70b415737645/mobly/test_runner.py#L407)):*
> ```Python
>                      records.OUTPUT_FILE_SUMMARY))
> 
>     # When a SIGTERM is received during the execution of a test, the Mobly test
> ```

clarify the reasoning for the location of this and the signal being used

**Review by booneng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/858)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @booneng, @mhaoli, and @xpconanfan)

___
*[`mobly/test_runner.py` line 407 at r5](https://reviewable.io/reviews/google/mobly/858#-NJ3wspm3FL4ZZVNAWla:-NJ4DtrD7PbxnfwOsVi9:b-896fix) ([raw file](https://github.com/google/mobly/blob/7e7cfdc6a5706b65a620c9bcc44a70b415737645/mobly/test_runner.py#L407)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

clarify the reasoning for the location of this and the signal being used
</blockquote></details>

Done.

**Review by xpconanfan — APPROVED** [review]

**Pull request merged**